// @sap-ux/preview-middleware/dist/client/cpe/outline/nodes.js
"use strict";

sap.ui.define(["./utils"], function (___utils) {
  "use strict";

  const isEditable = ___utils["isEditable"];
  function getAdditionalData(id) {
    const control = sap.ui.getCore().byId(id);
    if (control?.getMetadata().getProperty('text')) {
      const text = control.getProperty('text');
      if (typeof text === 'string' && text.trim() !== '') {
        return {
          text
        };
      }
    }
    return {};
  }

  /**
   * Transform node.
   *
   * @param input outline view node
   * @param getAdditionalData gets additional data for give control id.
   * @returns Promise<OutlineNode[]>
   */
  async function transformNodes(input) {
    const stack = [...input];
    const items = [];
    while (stack.length) {
      const current = stack.shift();
      const editable = await isEditable(current?.id);
      if (current?.type === 'element') {
        const children = (current.elements ?? []).flatMap(element => element.type === 'aggregation' ? element.elements ?? [] : []);
        const {
          text
        } = getAdditionalData(current.id);
        const node = {
          controlId: current.id,
          controlType: current.technicalName,
          name: text ?? current.technicalName.split('.').slice(-1)[0],
          editable,
          visible: current.visible ?? true,
          children: await transformNodes(children)
        };
        items.push(node);
      }
    }
    return items;
  }
  var __exports = {
    __esModule: true
  };
  __exports.transformNodes = transformNodes;
  return __exports;
});
